### PR TITLE
Pp 26 search for unauth users

### PR DESCRIPTION
**Pull request by Tanyaa-a**
### Description
- Implemented `SearchPage` component:
  - Allows users to search for classes by title, description, and category.
  - Displays search results with pagination.
  - Fetches classes from the backend based on search criteria.

### Changes
- FilterContainer.jsx
- SearchBar.jsx
- SearchPage.jsx
- SearchResults.jsx

### Related issues
Responsive styles still in work
### Screenshots

Rendered results of all classes

Screen Shot 2024-08-15 at 9 02 30 PM

Search by Category/Title
Screen Shot 2024-08-15 at 9 02 46 PM

Search by Title
Screen Shot 2024-08-15 at 9 03 09 PM

Search by Description
Screen Shot 2024-08-15 at 9 03 27 PM

### CODE REVIEW

File: .env
Comment (vrudnitskaya) [inline .env:1]: Tanya, env file should not be pushed to the github. Please check the gitignore file.

File: src/components/layouts/Footer.jsx [hunk 1]
```diff
@@ -1,6 +1,6 @@
 const Footer = () => {
   return (
-    <footer className="flex w-screen  flex-row items-center text-center justify-center bg-lightGreen w-full h-24  text-black">
+    <footer className="flex w-screen  flex-row items-center text-center justify-center bg-lightGreen h-24  text-black">
```
Comment (vrudnitskaya) [inline src/components/layouts/Footer.jsx:3]: Please delete from classnames w-screen; add w-full after bg-lightGreen and change the height from h-24 to h-20

File: src/components/search/SearchResults.jsx [hunk 2]
```diff
@@ -1,8 +1,99 @@
...
-      <h1>SearchResults</h1>
-    </>
+    <div className="container m-auto ">
+      {classes.length > 0 ? (
+        <>
+          <div className="flex flex-col gap-6 mt-10 ">
+            {classes.map((classItem) => (
+              <div
+                key={classItem._id}
+                className="flex flex-row border h-[300px] rounded-lg  border-gray hover:shadow-lg cursor-pointer "
```
Comment (vrudnitskaya) [inline src/components/search/SearchResults.jsx:20]: maybe its better to add w-11/12 self-center and change height h-1/4

File: src/components/search/SearchResults.jsx [hunk 3]
```diff
@@ -1,8 +1,99 @@
...
+import React from 'react';
+import { useNavigate } from 'react-router-dom';
+import AgeIcon from '../../assets/icons/icon-age.svg';
+import LessonTypeIcon from '../../assets/icons/icon-lesson.svg';
+import ScheduleIcon from '../../assets/icons/icons-schedule.svg';
+
+const SearchResults = ({ classes, currentPage, totalPages, onPageChange }) => {
+  const navigate = useNavigate();
+  const handleClick = () => {
+    navigate('/register');
```
Comment (vrudnitskaya) [inline src/components/search/SearchResults.jsx:10]: I think we need navigate user to login page.

File: .env
Comment (Tanyaa-a) [reply .env:1]: Thank you Valya! Removed!

File: src/components/search/SearchResults.jsx (hunk 2 above)
Comment (Tanyaa-a) [reply src/components/search/SearchResults.jsx:20]: still  working on styles.  Will remove hardcoded values while doing responsive

File: src/components/layouts/Footer.jsx (hunk 1 above)
Comment (Tanyaa-a) [reply src/components/layouts/Footer.jsx]: I did that and now I have narrow footer line is this ok?

File: src/components/layouts/Header.jsx [hunk 4]
```diff
@@ -1,33 +1,54 @@
...
 
-    return (
-        <header className="h-20 w-full flex items-center px-4 lg:px-20 justify-between">
-            <Link className="text-black font-spartan font-semibold text-xl sm:text-3xl w-1/3" to="/">TalentStudio</Link>
-            <DesktopNav onSearch={search} onLogin={login} onJoin={register} isSearch={isSearch} isLogin={isLoginBtn} isJoin={isJoinBtn}/>
-            <MobileNav onSearch={search} onLogin={login} onJoin={register} isSearch={isSearch} isLogin={isLoginBtn} isJoin={isJoinBtn}/>
-        </header>
-    );
+  const navigate = useNavigate();
+  const search = () => navigate('/search');
```
Comment (AnnaPestova1) [inline src/components/layouts/Header.jsx:22]: This logic just redirect user on the search page, but does not save search request. Tat's why the user need to enter search request again after redirection on the separate search bar. It is not use friendly. I think you have at least 2 approaches:
1. show search results after redirection without printing them again in a separate input
2. get rid of search input in the Navbar if this input does not handle search requests.

Please think about what to do in this case and make sure that you have ticket for this in the Jira

File: src/components/search/FilterContainer.jsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import iconsArrowDown from '../../assets/icons/icon.svg';
+import iconsSort from '../../assets/icons/icon-sort.svg';
+//filter functionality will work after backend will fin
```
Comment (AnnaPestova1) [inline src/components/search/FilterContainer.jsx:4]: make sure that you have backend and frontend tickets for this functionality

File: src/components/search/FilterContainer.jsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+const buttons = [
```
Comment (AnnaPestova1) [inline src/components/search/FilterContainer.jsx:5]: I like how you create the array for buttons and use map method to render the button.

File: src/components/search/SearchBar.jsx [hunk 7]
```diff
@@ -1,8 +1,53 @@
...
+          type="button"
+          onClick={handleClear}
+          aria-label="Clear search input"
+        >
+          <img src={deleteSearchIcon} className="" alt="clear search icon" />
+        </button>
+        <button
+          type="submit"
+          className="bg-darkGreen text-white font-roboto text-lg sm:text-xl md:text-2xl lg:text-3xl w-2/5 sm:w-1/2 md:w-1/3 lg:w-1/6 rounded-full"
+          aria-label="Search"
```
Comment (AnnaPestova1) [inline src/components/search/SearchBar.jsx:45]: here  the area-label is redundant: [By default, a button's accessible name is the content between the opening and closing [<button>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button) tags, an image's accessible name is the content of its [alt](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#alt) attribute, and a form input's accessible name is the content of the associated [<label>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label) element.

If none of these options are available or if the default accessible name is not appropriate, use the aria-label attribute to define the accessible name of an element.](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label)

File: src/components/search/SearchPage.jsx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  useEffect(() => {
+    fetchClasses(search, currentPage);
+  }, [search, currentPage, category]);
+
+  const handleSearch = (searchTerm) => {
+    setSearch(searchTerm);
+    setCurrentPage(1);
+  };
+
+  const handlePageChange = (page) => {
```
Comment (AnnaPestova1) [inline src/components/search/SearchPage.jsx:47]: minor comment: it is a good pattern to hide pagination button if result is only one page

File: src/components/search/SearchPage.jsx [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+import FilterButtons from './FilterContainer';
+
+const SearchPage = () => {
+  const [classes, setClasses] = useState([]);
+  const [currentPage, setCurrentPage] = useState(1);
+  const [totalPages, setTotalPages] = useState(1);
+  const [search, setSearch] = useState('');
+  const [category, setCategory] = useState('');
+
+  const fetchClasses = async (searchTerm = '', page = 1) => {
```
Comment (AnnaPestova1) [inline src/components/search/SearchPage.jsx:14]: you can move this function in a separate file

File: src/components/search/SearchPage.jsx [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+      setTotalPages(data.totalPages || 1);
+      setCurrentPage(data.currentPage || 1);
+    } catch (error) {
+      console.error('Failed to fetch classes:', error.message);
+    }
+  };
+
+  useEffect(() => {
+    fetchClasses(search, currentPage);
+  }, [search, currentPage, category]);
```
Comment (AnnaPestova1) [inline src/components/search/SearchPage.jsx:40]: do we need category dependency in this useEffect?

File: src/components/search/SearchResults.jsx [hunk 11]
```diff
@@ -1,8 +1,102 @@
-const SearchResults = () => {
+import React from 'react';
+import { useNavigate } from 'react-router-dom';
+import AgeIcon from '../../assets/icons/icon-age.svg';
+import LessonTypeIcon from '../../assets/icons/icon-lesson.svg';
+import ScheduleIcon from '../../assets/icons/icons-schedule.svg';
+
+const SearchResults = ({ classes, currentPage, totalPages, onPageChange }) => {
+  const navigate = useNavigate();
+  const handleClick = () => {
```
Comment (AnnaPestova1) [inline src/components/search/SearchResults.jsx:9]: this function always redirect to the /register page, even if user already logged in, I see that it probably out of scope of this user but make sure you have ticket for show class infirmation for registered users

Comment (AnnaPestova1) [inline src/components/search/SearchResults.jsx:9]: this function always redirect to the /register page, even if user already logged in, I see that it probably out of scope of this user but make sure you have ticket for show class infirmation for registered users

**Review by AnnaPestova1 — COMMENTED** [review]
Hi Tanya! Great functionality. I made some comments, make sure that you have the tickets created for all issues mentioned in comments.  

Handle please the Valentina's comments.

File: src/components/search/SearchBar.jsx (hunk 7 above)
Comment (Tanyaa-a) [reply src/components/search/SearchBar.jsx:45]: Thank  you! Got it!

File: src/components/search/SearchPage.jsx (hunk 9 above)
Comment (Tanyaa-a) [reply src/components/search/SearchPage.jsx:14]: Good idea! Thank you!

File: src/components/search/SearchResults.jsx (hunk 11 above)
Comment (Tanyaa-a) [reply src/components/search/SearchResults.jsx:9]: Thank you! I got it. Will change it to conditionally navigate depending on whether the user is logged in or not

File: src/components/search/SearchResults.jsx (hunk 3 above)
Comment (Tanyaa-a) [reply src/components/search/SearchResults.jsx:10]: Thank you Valya! Got it

File: src/components/search/SearchPage.jsx (hunk 10 above)
Comment (Tanyaa-a) [reply src/components/search/SearchPage.jsx:40]: Both. As for now we search classes by title, description and category they are in

File: src/components/layouts/Header.jsx (hunk 4 above)
Comment (vrudnitskaya) [reply src/components/layouts/Header.jsx:22]: I think Tanya needs to add a state for search input in the header, store the data via onChange and then when the user click on the button or press the key Enter, the state will be passed as the initial state of the search input on the Search page.

Comment (AnnaPestova1) [reply src/components/layouts/Header.jsx:22]: > I think Tanya needs to add a state for search input in the header, store the data via onChange and then when the user click on the button or press the key Enter, the state will be passed as to the search input on the Search page.

We discussed it with Tanya and decided that it is the next ticket.

**Review by AnnaPestova1 — APPROVED** [review]
Good job Tanya! Be sure that you'll handle search through search input in the header in upcoming ticket.

**Review by vrudnitskaya — APPROVED** [review]

**Pull request merged**